File: commonplace/format.py
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import List

from commonplace.instantiate import generate_instances_of_moment
from commonplace.models import (DocPosition, Moment, Outline, RecurringMoment, SingleMoment, Todos, WorkState)
from commonplace.util import get_bottom_line, with_start_of_day

TODO_FORMAT = "todo"
TRASH_FORMAT = "trash"
CAT_STYLE = "cat"
MOM_STYLE = "mom"
COM_STYLE = "com"
DATE_STYLE = "date"
TIME_STYLE = "time"
DONE_SUFFIX = ".done"
PRIORITY_SUFFIX = ".priority"
UNTIL_SUFFIX = ".until"


@dataclass
class FormatState:
    raw_content: str
    format: str = ""
    last_style: str = ""
    last_style_start: int = 0
    last_style_end: int = 0
    fixed_time: datetime | None = None


def format_todos(todos: Todos, raw_content: str, format_type: str = TODO_FORMAT, fixed_time=None) -> str:
    state = FormatState(raw_content=raw_content, fixed_time=fixed_time)

    for cat in todos.categories:
        add_format_line(state, CAT_STYLE, cat.doc_pos)

    for mom in todos.moments:
        if format_type == TODO_FORMAT:
            format_moment(state, mom)
        elif format_type == TRASH_FORMAT:
            format_trash_moment(state, mom)
        else:
            raise ValueError(f"Unknown format '{format_type}'")

    if state.last_style:
        flush_last_style(state)

    return state.format


def format_moment(state: FormatState, mom: Moment, parent_done=False):
    style = MOM_STYLE
    done = parent_done or mom.work_state == WorkState.DONE
    if done:
        style += DONE_SUFFIX
    else:
        style += format_due_soon(mom, fixed_time=state.fixed_time)
        if mom.priority > 0:
            style += PRIORITY_SUFFIX

    add_format_line(state, style, mom.doc_pos)

    if done:
        for com in mom.comments:
            add_format_line(state, COM_STYLE + DONE_SUFFIX, com.doc_pos)
    else:
        format_dates(state, mom)

    for sub in mom.sub_moments:
        format_moment(state, sub, parent_done=done)


def format_trash_moment(state: FormatState, mom: Moment):
    add_format_line(state, MOM_STYLE, mom.doc_pos)
    format_dates(state, mom)

    for sub in mom.sub_moments:
        format_trash_moment(state, sub)


def format_dates(state: FormatState, mom: Moment):
    if isinstance(mom, SingleMoment):
        if mom.start:
            add_format_line(state, DATE_STYLE, mom.start.doc_pos)
        if mom.end and (not mom.start or mom.end.doc_pos != mom.start.doc_pos):
            add_format_line(state, DATE_STYLE, mom.end.doc_pos)

    elif isinstance(mom, RecurringMoment):
        add_format_line(state, DATE_STYLE, mom.recurrence.ref_date.doc_pos)

    if mom.time_of_day:
        add_format_line(state, TIME_STYLE, mom.time_of_day.doc_pos)


def format_due_soon(mom: Moment, fixed_time=None):
    # Due until 10 (n-1) days in the future
    cutoff = 11
    today = with_start_of_day(fixed_time if fixed_time else datetime.now()).astimezone(tz=None)
    n_days_from_today = today + timedelta(days=cutoff)
    n_real_hours = (n_days_from_today - today) / timedelta(hours=1)
    instances = generate_instances_of_moment(mom, today, n_days_from_today, incl_subs=False)
    earliest = cutoff
    for inst in instances:
        due = inst.end - today
        # We need to compare hours here because of daylight saving time.
        # Instead of 264h (=11 days) it might only be 263h or 265h,
        # which would lead to the wrong number of days calculated.
        if due / timedelta(hours=1) >= n_real_hours:
            continue

        due_days = int(due / timedelta(days=1))
        if due_days < earliest:
            earliest = due_days

    if earliest < cutoff:
        return UNTIL_SUFFIX + str(earliest)

    return ""


def add_format_line(state: FormatState, style: str, pos: DocPosition):
    if state.last_style and style == state.last_style and only_whitespace_between(state.last_style_end, pos.offset,
                                                                                  state.raw_content):
        state.last_style_end = pos.offset + pos.length
    else:
        if state.last_style:
            flush_last_style(state)
        state.last_style = style
        state.last_style_start = pos.offset
        state.last_style_end = pos.offset + pos.length


def flush_last_style(state: FormatState):
    state.format += f"{state.last_style_start},{state.last_style_end},{state.last_style}\n"


def only_whitespace_between(start, end, raw_content):
    return all((c.isspace() for c in raw_content[start:end]))


def fold_todos(todos: Todos) -> str:
    folds = ""
    for mom in todos.moments:
        start = mom.doc_pos.line_num
        end = get_bottom_line(mom)
        if end > start:
            folds += f"{start}-{end}\n"

    return folds


def outline_todos(todos: Todos, raw_content: str, format_type: str = TODO_FORMAT) -> List[Outline]:
    return [outline(mom, raw_content) for mom in todos.moments if format_type != TODO_FORMAT or mom.work_state != WorkState.DONE]


def outline(mom: Moment, raw_content: str) -> Outline:
    detail = ""
    if isinstance(mom, SingleMoment):
        if mom.start:
            detail += extract(raw_content, mom.start.doc_pos)
        if mom.end and (not mom.start or mom.end.doc_pos != mom.start.doc_pos):
            detail += " - " + extract(raw_content, mom.end.doc_pos)

    elif isinstance(mom, RecurringMoment):
        detail += extract(raw_content, mom.recurrence.ref_date.doc_pos)

    if mom.time_of_day:
        detail += " " + extract(raw_content, mom.time_of_day.doc_pos)

    return Outline(
        name=mom.name,
        start_line=mom.doc_pos.line_num,
        end_line=get_bottom_line(mom),
        detail=detail,
    )


def extract(content: str, doc_pos: DocPosition) -> str:
    return content[doc_pos.offset:doc_pos.offset + doc_pos.length]

File: commonplace_js/packages/lib/format.ts
import { addDays, differenceInDays, differenceInHours, startOfDay } from "date-fns"
import { generateInstancesOfMoment } from "./instantiate"
import { Moment, Todos, WorkState, SingleMoment, isSingleMoment, isRecurringMoment, RecurringMoment, DocPosition, Outline } from "./models"
import { getBottomLine } from "./util"

export const TODO_FORMAT = 'todo'
export const TRASH_FORMAT = "trash"
const CAT_STYLE = "cat"
const MOM_STYLE = "mom"
const COM_STYLE = "com"
const DATE_STYLE = "date"
const TIME_STYLE = "time"
const DONE_SUFFIX = ".done"
const PRIORITY_SUFFIX = ".priority"
const UNTIL_SUFFIX = ".until"

interface FormatState {
    rawContent: string
    styles: FormatStyle[]
    lastStyle?: FormatStyle
    fixedTime: Date | null
}

export interface FormatStyle {
    start: number
    end: number
    style: string
}

export function formatTodos(todos: Todos, rawContent: string, formatType: string = TODO_FORMAT, fixedTime: Date | null = null): FormatStyle[] {
    const state: FormatState = { rawContent, fixedTime, styles: [] }

    for (const cat of todos.categories) {
        addFormatLine(state, CAT_STYLE, cat.docPos)
    }

    for (const mom of todos.moments) {
        if (formatType === TODO_FORMAT) {
            formatMoment(state, mom)
        }
        else if (formatType === TRASH_FORMAT) {
            formatTrashMoment(state, mom)
        }
        else {
            throw Error(`Unknown format '{formatType}'`)
        }
    }

    if (state.lastStyle) {
        flushLastStyle(state)
    }

    return state.styles
}

function formatMoment(state: FormatState, mom: Moment, parentDone = false) {
    let style = MOM_STYLE
    const done = parentDone || mom.workState === WorkState.DONE
    if (done) {
        style += DONE_SUFFIX
    }
    else {
        style += formatDueSoon(mom, state.fixedTime)
        if (mom.priority > 0) {
            style += PRIORITY_SUFFIX
        }
    }

    addFormatLine(state, style, mom.docPos)

    if (done) {
        for (const com of mom.comments) {
            addFormatLine(state, COM_STYLE + DONE_SUFFIX, com.docPos)
        }
    }
    else {
        formatDates(state, mom)
    }

    for (const sub of mom.subMoments) {
        formatMoment(state, sub, done)
    }
}

function formatTrashMoment(state: FormatState, mom: Moment) {
    addFormatLine(state, MOM_STYLE, mom.docPos)
    formatDates(state, mom)

    for (const sub of mom.subMoments) {
        formatTrashMoment(state, sub)
    }
}

function formatDates(state: FormatState, mom: Moment) {
    if (isSingleMoment(mom)) {
        const singleMom = mom as SingleMoment;
        if (singleMom.start) {
            addFormatLine(state, DATE_STYLE, singleMom.start.docPos!)
        }
        if (singleMom.end && (!singleMom.start || singleMom.end.docPos !== singleMom.start.docPos)) {
            addFormatLine(state, DATE_STYLE, singleMom.end.docPos!)
        }
    }
    else if (isRecurringMoment(mom)) {
        addFormatLine(state, DATE_STYLE, (mom as RecurringMoment).recurrence.refDate.docPos!)
    }

    if (mom.timeOfDay) {
        addFormatLine(state, TIME_STYLE, mom.timeOfDay.docPos!)
    }
}

function formatDueSoon(mom: Moment, fixedTime: Date | null) {
    // Due until 10 (n-1) days in the future
    const cutoff = 11
    const today = startOfDay(fixedTime ? fixedTime : new Date())
    const nDaysFromToday = addDays(today, cutoff)
    const nRealHours = differenceInHours(nDaysFromToday, today)
    const instances = generateInstancesOfMoment(mom, today, nDaysFromToday, false)
    let earliest = cutoff
    for (const inst of instances) {
        // We need to compare hours here because of daylight saving time.
        // Instead of 264h (=11 days) it might only be 263h or 265h,
        // which would lead to the wrong number of days calculated.
        if (differenceInHours(inst.end, today) >= nRealHours) {
            continue
        }

        const dueDays = differenceInDays(inst.end, today)
        if (dueDays < earliest) {
            earliest = dueDays
        }
    }

    if (earliest < cutoff) {
        return `${UNTIL_SUFFIX}${earliest}`
    }

    return ""
}

function addFormatLine(state: FormatState, style: string, pos: DocPosition) {
    if (state.lastStyle && style === state.lastStyle.style && onlyWhitespaceBetween(state.lastStyle.end, pos.offset, state.rawContent)) {
        state.lastStyle.end = pos.offset + pos.length
    }
    else {
        if (state.lastStyle) {
            flushLastStyle(state)
        }
        state.lastStyle = {
            start: pos.offset,
            end: pos.offset + pos.length,
            style: style
        }
    }
}

function flushLastStyle(state: FormatState) {
    state.styles.push(state.lastStyle!);
    state.lastStyle = undefined;
}

function onlyWhitespaceBetween(start: number, end: number, rawContent: string) {
    for (let i = start; i < end; i++) {
        if (!/\s/.test(rawContent[i])) {
            return false;
        }
    }
    return true;
}

export function foldTodos(todos: Todos): number[][] {
    return todos.moments
        .map(mom => [mom.docPos.lineNum, getBottomLine(mom)])
        .filter(([start, end]) => end > start);
}


export function outlineTodos(todos: Todos, raw_content: string, format_type = TODO_FORMAT): Outline[] {
    return todos.moments
        .filter(mom => format_type !== TODO_FORMAT || mom.workState !== WorkState.DONE)
        .map(mom => outline(mom, raw_content));
}


function outline(mom: Moment, raw_content: string): Outline {
    let detail = ''
    if (isSingleMoment(mom)) {
        const singleMom = mom as SingleMoment;
        if (singleMom.start) {
            detail += extract(raw_content, singleMom.start.docPos!)
        }
        if (singleMom.end && (!singleMom.start || singleMom.end.docPos?.offset !== singleMom.start.docPos?.offset)) {
            detail += " - " + extract(raw_content, singleMom.end.docPos!)
        }
    }
    else if (isRecurringMoment(mom)) {
        detail += extract(raw_content, (mom as RecurringMoment).recurrence.refDate.docPos!)
    }

    if (mom.timeOfDay) {
        detail += " " + extract(raw_content, mom.timeOfDay.docPos!)
    }

    return {
        name: mom.name,
        startLine: mom.docPos.lineNum,
        endLine: getBottomLine(mom),
        detail: detail,
    }
}

function extract(content: string, docPos: DocPosition): string {
    return content.slice(docPos.offset, docPos.offset + docPos.length)
}